Rename GH char speeds, minor simple optimizations in GH DuDt

nilsdeppe opened the pull request:
## Proposed changes

- Rename GH char speeds from `u` -> `v` in the code
- Add some more temp tags to `Variables.hpp`
- Group together allocations in `ComputeDuDt` for GH
- Avoid zero-initialization in `ComputeDuDt` for GH
- Avoid recomputing storage index in some common cases for `ComputeDuDt` for GH

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Rename GH char fields from UName to VName

Also update comments in GH/Characterstics.cpp to denote speeds with
lambda(CHAR_FIELD)

nilsdeppe committed:
Add more Temp tensor tags

nilsdeppe committed:
Eliminate allocations in GH ComputeDuDt

nilsdeppe committed:
Optimize generalized harmonic ComputeDuDt

markscheel commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -150,7 +181,8 @@ void ComputeDuDt<Dim>::apply(
   }
 
-  tnsr::abC<DataVector, Dim> phi_3_up{DataVector(n_pts, 0.)};
+  tnsr::iaB<DataVector, Dim>& phi_3_up =
```
Ugh, I didn't know that previously we allocated more space than we used.

markscheel reviewed (commented):
Looks good. 

For the 'Rename GH char speeds...' commit:  
 - It renames the char fields as well as the char speeds, so you might want to reword the commit message appropriately.  
 - The char fields are named 'U' in all the papers so why the change to 'V'?

nilsdeppe commented:
@markscheel thanks for the review! I'll update the commit message :) 

As per the char field names: we discussed this on a scalar wave PR with @kidder and the issue is that there is no cross-field convention. SpEC vacuum uses `u`, a lot of hydro and DG papers uses `w`, just as many use `v`, others use `q`, and yet others use `Q`. The goal is to use one convention throughout SpECTRE so at least our code (and hopefully papers!) will be consistent. We settled on `u_\alpha` for the collection of evolved fields (this is reasonably standard), `v_\hat{\alpha}` for the characteristic fields associated with `u_\alpha`, `p_\alpha` for the primitive variables, and `q_\hat{\alpha}` for the characteristic fields associated with `p_\alpha`.

markscheel commented:
Ok about the names, consistency is a good thing.

nilsdeppe commented:
Okay, rebased on develop, update the commit name, and added a fixup that changes (I'll squash it into the char field rename) `v(CHAR_FIELD)` to `lambda(CHAR_FIELD)` since that's what we decided to go with for the speeds

markscheel reviewed (commented):
Looks good; pls squash.

nilsdeppe commented:
Squashed and rebased on develop. Thanks for the review @markscheel!

moxcodes commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -280,8 +280,10 @@ void ComputeDuDt<Dim>::apply(
-            d_spacetime_metric.get(n, mu, nu) - phi.get(n, mu, nu);
+        const size_t storage_index =
+            three_index_constraint.get_storage_index(n, mu, nu);
+        three_index_constraint[storage_index] =
```
The two-step indexing is a little cumbersome, especially if we're going to start peppering it in all of the performance-sensitive tensor calculations. Any chance we could make the run-time `get`s into a lookup into a compile-time table if we want the index computations to have the extra optimization?

nilsdeppe replied:
They already do this, and it's that index collapsing that can end up being a bit slow. However, I'll drop the commit from this PR and stick it on a branch. Once we are at a point where things work we can revisit this.

moxcodes replied:
alright, sounds good -- it would be nice to get the index access well-optimized, but I'm hopeful for a general solution so we don't have to do it manually everywhere. :)

nilsdeppe replied:
Agreed :) Full-blown tensor expressions should be able to do that since they would be able to generate the for loops, and hence the storage indices at compile time :)

moxcodes reviewed (commented):
I'm happy with everything except the index caching, which just seems a little tiresome to apply widely.

nilsdeppe commented:
Dropped the commit and rebased on the latest develop. Thanks for the reviews @markscheel and @moxcodes!

kidder commented:
ping @markscheel
